File: 23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */
#include <vector>

using std::vector;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    // Linked list + Divide and conquer + two pointers
    // T: O(n * log(k)), k := the number of lists, n := avg. length of list
    // S: O(log(k)) if using recursive partition, O(1) if using iterative partition
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // Recursive partition
        // return partition(lists, 0, lists.size() - 1);

        // Iterative partition
        if (lists.empty()) {
            return nullptr;
        }

        int n = lists.size();
        while (n > 1) {
            for (int l = 0; l < (n >> 1); ++l) {
                int r = n - 1 - l;
                lists[l] = mergeTwoLists(lists[l], lists[r]);
            }
            n = (n + 1) >> 1;
        }
        return lists.front();
    }

private:
    // S: O(log(k))
    static ListNode* partition(vector<ListNode*>& lists, int l, int r) {
        if (l == r) {
            return lists[l];   // the number of linked lists is odd
        } else if (l > r) { 
            return nullptr;
        } else {
            int m = l + ((r - l) >> 1);
            ListNode* l1 = partition(lists, l, m);
            ListNode* l2 = partition(lists, m + 1, r);
            return mergeTwoLists(l1, l2);
        }
    }

    /*
    // Recursively merges two lists (might be faster in some languages), 
    // and avoids creating a dummy node (but recursion stack exists)
    // S: O(n)
    static ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr) {
            return l2;
        }
        if (l2 == nullptr) {
            return l1;
        }
        if (l1->val < l2->val) {
            l1->next = mergeTwoLists(l1->next, l2);
            return l1;
        } else {
            l2->next = mergeTwoLists(l1, l2->next);
            return l2;
        }
    }
    */
 
    // Iteratively merges two lists, like Q21
    // S: O(1)
    static ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode();
        ListNode* cur = dummy;
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        while (cur1 != nullptr && cur2 != nullptr) {
            if (cur1->val < cur2->val) {
                cur->next = cur1;
                cur1 = cur1->next;
            } else {
                cur->next = cur2;
                cur2 = cur2->next;
            }
            cur = cur->next;
        }

        cur->next = cur1 != nullptr ? cur1 : cur2;

        ListNode* head = dummy->next;
        delete dummy;
        return head;
    }
};
// @lc code=end

File: 23.merge-k-sorted-lists.go
/*
 * @lc app=leetcode id=23 lang=golang
 *
 * [23] Merge k Sorted Lists
 */

type ListNode struct {
	Val int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// T: O(len(list1) + len(list2))
func conquer(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}
	
	dummy := &ListNode{}
	cur := dummy
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			cur.Next = list1
			list1 = list1.Next
		} else {
			cur.Next = list2
			list2 = list2.Next
		}
		cur = cur.Next
	}

	if list1 != nil {
		cur.Next = list1
	} else {
		cur.Next = list2
	}

	return dummy.Next
}


func divideAndConquer(lists []*ListNode, l int, r int) *ListNode {
	if l > r {
		return nil
	} else if l == r {
		return lists[l]
	} else {
		m := l + ((r - l) >> 1)
		list1 := divideAndConquer(lists, l, m)
		list2 := divideAndConquer(lists, m + 1, r)
		return conquer(list1, list2)
	}
}


// T: O(m * log(n)), m := max length of one list, n := the number of lists
// S: O(log(n)) recursion stack
func mergeKLists(lists []*ListNode) *ListNode {
    return divideAndConquer(lists, 0, len(lists) - 1)
}
// @lc code=end

